train_scripts/train_sana_sprint_diffusers.py

```diff
@@ -0,0 +1,1823 @@
+        transformer.train()
+        disc.train()
+
+        for step, batch in enumerate(train_dataloader):
+            # text encoding
+            prompts = batch["prompts"]
+            text_encoding_pipeline = text_encoding_pipeline.to("cuda")
+            with torch.no_grad():
+                prompt_embeds, prompt_attention_mask, _, _ = text_encoding_pipeline.encode_prompt(
+                    prompts, complex_human_instruction=complex_human_instruction, do_classifier_free_guidance=False
+                )
+
+            # Convert images to latent space
+            vae = vae.to(accelerator.device)
+            pixel_values = batch["pixel_values"].to(dtype=vae.dtype)
+            model_input = vae.encode(pixel_values).latent
+            model_input = model_input * vae_config_scaling_factor * sigma_data
+            model_input = model_input.to(dtype=weight_dtype)
+
+            # Sample noise that we'll add to the latents
+            noise = torch.randn_like(model_input) * sigma_data
+            bsz = model_input.shape[0]
+
+            # Sample a random timestep for each image
+            # for weighting schemes where we sample timesteps non-uniformly
+            u = compute_density_for_timestep_sampling_scm(
+                batch_size=bsz,
+                logit_mean=args.logit_mean,
+                logit_std=args.logit_std,
+            )
+
+            # Add noise according to TrigFlow.
+            # zt = cos(t) * x + sin(t) * noise
+            t = u.view(-1, 1, 1, 1)
+            noisy_model_input = torch.cos(t) * model_input + torch.sin(t) * noise
+
+            scm_cfg_scale = torch.tensor(
+                np.random.choice(args.scm_cfg_scale, size=bsz, replace=True),
+                device=accelerator.device,
+            )
+
+            def model_wrapper(scaled_x_t, t):
+                pred = accelerator.unwrap_model(transformer)(
+                    hidden_states=scaled_x_t,
+                    timestep=t.flatten(),
+                    encoder_hidden_states=prompt_embeds,
+                    encoder_attention_mask=prompt_attention_mask,
+                    guidance=(scm_cfg_scale * args.guidance_embeds_scale),
+                    jvp=True,
+                    return_logvar=True,
+                )[0]
+                return pred
+
+            if phase == "G":
+                transformer.train()
+                disc.eval()
+                models_to_accumulate = [transformer]
+                with accelerator.accumulate(models_to_accumulate):
+                    with torch.no_grad():
+                        cfg_x_t = torch.cat([noisy_model_input, noisy_model_input], dim=0)
+                        cfg_t = torch.cat([t, t], dim=0)
+                        cfg_y = torch.cat([negative_prompt_embeds, prompt_embeds], dim=0)
+                        cfg_y_mask = torch.cat([negative_prompt_attention_mask, prompt_attention_mask], dim=0)
+
+                        cfg_pretrain_pred = pretrained_model(
```

Comment:
As another optimization, we could keep the `pretrained_model` in CPU once this computation is done and load to GPU again when needed.